File: src/features/typing/data/java/coin_change.java
import java.util.Arrays;

class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = Integer.MAX_VALUE;
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, max);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (coin <= i && dp[i - coin] != max && dp[i - coin] + 1 < dp[i]) {
                    dp[i] = dp[i - coin] + 1;
                }
            }
        }
        if (dp[amount] == max) {
            return -1;
        }
        return dp[amount];
    }
}

File: src/features/typing/data/ts/coin_change.ts
function coinChange(coins: number[], amount: number): number {
    const max = Number.MAX_SAFE_INTEGER;
    const dp: number[] = new Array(amount + 1).fill(max);
    dp[0] = 0;
    for (let i = 1; i <= amount; i++) {
        for (const coin of coins) {
            if (coin <= i && dp[i - coin] !== max && dp[i - coin] + 1 < dp[i]) {
                dp[i] = dp[i - coin] + 1;
            }
        }
    }
    if (dp[amount] === max) {
        return -1;
    }
    return dp[amount];
}
